helix-core/src/main/java/org/apache/helix/controller/changedetector/trimmer/HelixPropertyTrimmer.java

```diff
@@ -57,38 +57,49 @@
    * @param originalProperty
    */
   protected ZNRecord doTrim(T originalProperty) {
+    ZNRecord originalZNRecord = originalProperty.getRecord();
     ZNRecord trimmedZNRecord = new ZNRecord(originalProperty.getId());
-    for (Map.Entry<FieldType, Set<String>> fieldEntry : getNonTrimmableFields(
-        originalProperty).entrySet()) {
+    Map<FieldType, Set<String>> nonTrimmableFields = getNonTrimmableFields(originalProperty);
+
+    // Ensure the keys of all map fields and list fields are preserved even after the trim.
+    // The keys of list and map fields are relatively stable and contain important information. So
+    // they should not be trimmed.
+    originalZNRecord.getMapFields().keySet().stream()
+        .forEach(key -> trimmedZNRecord.setMapField(key, Collections.EMPTY_MAP));
+    originalZNRecord.getListFields().keySet().stream()
+        .forEach(key -> trimmedZNRecord.setListField(key, Collections.EMPTY_LIST));
```

Comment:
This part of code may come as a surprise to future developers that extend this class. When `getNonTrimmableFields()` is provided for overriding, it's natural to think `getNonTrimmableFields()` provides full control on "what fields to keep". This logic here, while necessary, may become debug overhead. It's also hard to change this behavior in the future if it stays in `doTrim()` like so. 

I suggest creating another function such as `getNonTrimmableKeysOnly()` and implement it under `HelixPropertyTrimmer.java` to preserve MapField and ListField keys by default; in `doTrim()`, we call that function and preserve keys only. The extended classes may choose to override it (and for now, they shouldn't). This is a clearer approach to creating this abstract class.

Reply:
Good idea. It helps to clean up the code.